tests/db.rs

```diff
@@ -0,0 +1,184 @@
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use slatedb::config::{CompactorOptions, DbOptions, PutOptions, WriteOptions};
+use slatedb::object_store::memory::InMemory;
+use slatedb::object_store::path::Path;
+use slatedb::object_store::ObjectStore;
+use slatedb::Db;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 8)]
+async fn test_concurrent_writers_and_readers() {
+    const NUM_WRITERS: usize = 10;
+    const NUM_READERS: usize = 2;
+    const WRITES_PER_TASK: usize = 100;
+    const KEY_LENGTH: usize = 256;
+
+    // Pad keys to allow us to control how many blocks we take up
+    // Since block size is not configurable
+    fn zero_pad_key(key: u64, length: usize) -> Vec<u8> {
+        let mut padded_key = vec![0; length];
+        // Set the first 8 bytes to the key
+        padded_key[0..8].copy_from_slice(&key.to_le_bytes());
+        padded_key
+    }
+
+    // Create an InMemory object store and DB
+    let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+    let config = DbOptions {
+        flush_interval: Some(Duration::from_millis(100)),
+        manifest_poll_interval: Duration::from_millis(100),
+        compactor_options: Some(CompactorOptions {
+            poll_interval: Duration::from_millis(100),
+            ..Default::default()
+        }),
+        // Allow 16KB of unflushed data
+        max_unflushed_bytes: 16 * 1024,
+        min_filter_keys: 0,
+        // Allow up to four 4096-byte blocks per-SST
+        l0_sst_size_bytes: 4 * 4096,
+        ..Default::default()
+    };
+    let db = Arc::new(
+        Db::open_with_opts(
+            Path::from("/tmp/test_concurrent_writers_readers"),
+            config,
+            object_store.clone(),
+        )
+        .await
+        .unwrap(),
+    );
+
+    // Flag to signal readers to stop
+    let stop = Arc::new(AtomicBool::new(false));
+
+    // Writer tasks: each writer writes to its own key with incrementing values
+    let writer_handles = (0..NUM_WRITERS)
+        .map(|writer_id| {
+            let db = db.clone();
+            let stop_writers = stop.clone();
+            tokio::spawn(async move {
+                // Get a random key in the keyspace. Don't use nth. Pick one at random.
+                let key = zero_pad_key(writer_id.try_into().unwrap(), KEY_LENGTH);
```

Comment:
I didn't understand the comment.
This is always writing the same key for a writer.

Reply:
Oops! That's leftover 😅

Reply:
Fixed here: 4090b34